cache-config/testing/ort-tests/t3c-apply-diff_test.go

```diff
@@ -18,113 +18,107 @@ import (
+
+func writeComment(t *testing.T) {
+	f, err := os.OpenFile(recordsConfigFileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		t.Fatalf("opening file '%s': %v", recordsConfigFileName, err)
+	}
+	defer f.Close()
+	_, err = f.Write([]byte(" #mycomment\n"))
+	if err != nil {
+		t.Fatalf("writing comment to file '%s': %v", recordsConfigFileName, err)
+	}
+
+}
+
+func writeNonComment(t *testing.T) {
+	f, err := os.OpenFile(recordsConfigFileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		t.Fatalf("opening file '%s': %v", recordsConfigFileName, err)
+	}
+	_, err = f.Write([]byte("\nmynocomment this line isn't a comment\n"))
+	f.Close()
+	if err != nil {
+		t.Fatalf("writing line to file '%s': %v", recordsConfigFileName, err)
+	}
+}
+
+func verifyOverwritten(t *testing.T) {
+	recordsDotConfig, err := ioutil.ReadFile(recordsConfigFileName)
+	if err != nil {
+		t.Fatalf("reading %s: %v", recordsConfigFileName, err)
+	}
+	content := string(recordsDotConfig)
+	if strings.Contains(content, "#mycomment") {
+		t.Fatalf("expected records.config to have a diff and be replaced with a non-comment difference, actual: %s", content)
+	} else if strings.Contains(content, "mynocomment") {
+		t.Fatalf("expected records.config to have a diff and be replaced with a non-comment difference, actual: %s", content)
+	}
+}
+
 func TestApplyDiff(t *testing.T) {
-	testName := "TestApplyDiff"
-	t.Logf("------------- Starting " + testName + " ---------------")
 	tcd.WithObjs(t, []tcdata.TCObj{
 		tcdata.CDNs, tcdata.Types, tcdata.Tenants, tcdata.Parameters,
 		tcdata.Profiles, tcdata.ProfileParameters, tcdata.Statuses,
 		tcdata.Divisions, tcdata.Regions, tcdata.PhysLocations,
 		tcdata.CacheGroups, tcdata.Servers, tcdata.Topologies,
 		tcdata.DeliveryServices}, func() {
 
-		hostName := "atlanta-edge-03"
-		const fileName = `records.config`
-
 		// badass to get initial config files
 
-		if out, code := t3cUpdateReload(hostName, "badass"); code != 0 {
-			t.Fatalf("t3c apply badass failed: output '''%v''' code %v\n", out, code)
+		if out, code := t3cUpdateReload(cacheHostName, "badass"); code != 0 {
+			t.Fatalf("t3c apply badass failed with exit code %d, output: %s", code, out)
 		}
 
-		filePath := test_config_dir + "/" + fileName
-
-		if !util.FileExists(filePath) {
-			t.Fatalf("missing config file '%v' needed to test", filePath)
+		if !util.FileExists(recordsConfigFileName) {
+			t.Fatalf("missing config file '%s' needed to test", recordsConfigFileName)
 		}
 
-		// write a comment, which should diff as unchanged
-
-		{
-			f, err := os.OpenFile(filePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
-			if err != nil {
-				t.Fatalf("opening file '%v': %v", filePath, err)
-			}
-			_, err = f.Write([]byte(" #mycomment\n"))
-			f.Close()
-			if err != nil {
-				t.Fatalf("writing comment to file '%v': %v", filePath, err)
-			}
-		}
+		t.Run("write a comment, which should diff as unchanged", writeComment)
```

Comment:
I had a hard time understanding this. I typically think if `t.Run` as "tests," but it took me a while to figure out the functions here aren't tests themselves, which I wasn't expecting.

The things `writeComment` and `writeNonComment` are capable of failing, but they're not tests themselves, they're just setting up things to be tested.

The code between `t.Run(writeComment)` and `t.Run(writeNonComment)` is the real test. And then `verifyOverwritten` is the real test, but it doesn't include its own setup, that's in `writeNonComment`, which itself isn't a test but just setup.

What do you think about something like:

```
		// badass to get initial config files

		if out, code := t3cUpdateReload(cacheHostName, "badass"); code != 0 {
			t.Fatalf("t3c apply badass failed with exit code %d, output: %s", code, out)
		}
		if !util.FileExists(recordsConfigFileName) {
			t.Fatalf("missing config file '%s' needed to test", recordsConfigFileName)
		}

		t.Run("verify comment is unchanged", func(t *testing.T) {
			f, err := os.OpenFile(recordsConfigFileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
			if err != nil {
				t.Fatalf("opening file '%s': %v", recordsConfigFileName, err)
			}
			defer f.Close()
			_, err = f.Write([]byte(" #mycomment\n"))
			if err != nil {
				t.Fatalf("writing comment to file '%s': %v", recordsConfigFileName, err)
			}

			// queue and syncds to get changes

			if err := ExecTOUpdater(cacheHostName, false, true); err != nil {
				t.Fatalf("updating queue status failed: %v", err)
			}
			out, code := t3cUpdateReload(cacheHostName, "syncds")
			if code != 0 {
				t.Fatalf("t3c apply failed with exit code %d, output: %s", code, out)
			}

			// verify the file wasn't overwritten, as it would be if there were a diff

			recordsDotConfig, err := ioutil.ReadFile(recordsConfigFileName)
			if err != nil {
				t.Fatalf("reading %s: %v", recordsConfigFileName, err)
			}
			if !bytes.Contains(recordsDotConfig, []byte("#mycomment")) {
				t.Fatalf("expected records.config to diff clean and not be replaced with comment difference, actual: '%s' t3c-apply output: %s", string(recordsDotConfig), out)
			}
		})

		t.Run("verify non-comment is overwritten", func(t *testing.T) {
			f, err := os.OpenFile(recordsConfigFileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
			if err != nil {
				t.Fatalf("opening file '%s': %v", recordsConfigFileName, err)
			}
			_, err = f.Write([]byte("\nmynocomment this line isn't a comment\n"))
			f.Close()
			if err != nil {
				t.Fatalf("writing line to file '%s': %v", recordsConfigFileName, err)
			}

			// queue and syncds to get changes

			if err := ExecTOUpdater(cacheHostName, false, true); err != nil {
				t.Fatalf("updating queue status failed: %v", err)
			}
			out, code = t3cUpdateReload(cacheHostName, "syncds")
			if code != 0 {
				t.Fatalf("t3c apply failed with exit code %d, output: %s", code, out)
			}
			t.Logf("t3c apply output: %s", out)

			recordsDotConfig, err := ioutil.ReadFile(recordsConfigFileName)
			if err != nil {
				t.Fatalf("reading %s: %v", recordsConfigFileName, err)
			}
			content := string(recordsDotConfig)
			if strings.Contains(content, "#mycomment") {
				t.Fatalf("expected records.config to have a diff and be replaced with a non-comment difference, actual: %s", content)
			} else if strings.Contains(content, "mynocomment") {
				t.Fatalf("expected records.config to have a diff and be replaced with a non-comment difference, actual: %s", content)
			}
		})
```
?

Reply:
The grouping of `t.Run`s here was to reflect places where anonymous block scopes were being used. I don't have any problem restructuring to use two `t.Run`s that together group each scenario more logically, but I would want to put those in their own functions rather than anonymous function literals.

Reply:
I personally find the embedded functions a lot easier to read for tests like this, where you almost always want to see the tests together, and it almost never makes sense to view them in isolation, rather than having to jump back and forth to read the code. 

But readability is relative. Named functions would make me sad, but I wouldn't -1 it.

Reply:
`t.Run` calls with anonymous functions as their second argument are more easily followed than most, because of the necessary explanation in the first argument. In general, though, the deeper scopes become nested the more confusing a function becomes. I'll make the `t.Run` calls here inline, since there's no re-use with which to be concerned, and they aren't resulting in any insane levels of scope-nesting.